lnwallet/musig_session.go

```diff
@@ -0,0 +1,535 @@
+	// session, and their nonce the verification nonce.
+	case m.remoteCommit:
+		localNonce = m.nonces.SigningNonce
+		remoteNonce = m.nonces.VerificationNonce
+
+	// Otherwise, we're generating/receiving a signature for our local
+	// commitment (to broadcast), so now our verification nonce is the one
+	// we've already generated, and we want to bind their new signing
+	// nonce.
+	default:
+		localNonce = m.nonces.VerificationNonce
+		remoteNonce = m.nonces.SigningNonce
+	}
+
+	tweakDesc := input.MuSig2Tweaks{
+		TaprootBIP0086Tweak: true,
+	}
+	m.session, err = m.signer.MuSig2CreateSession(
+		m.localKey.KeyLocator, m.signerKeys, &tweakDesc,
+		[][musig2.PubNonceSize]byte{remoteNonce.PubNonce},
+		musig2.WithPreGeneratedNonce(&localNonce),
+	)
+	if err != nil {
+		return err
+	}
+
+	// We'll need the raw combined nonces later to be able to verify
+	// partial signatures, and also combine partial signatures, so we'll
+	// generate it now ourselves.
+	m.combinedNonce, err = musig2.AggregateNonces([][musig2.PubNonceSize]byte{
+		m.nonces.SigningNonce.PubNonce,
+		m.nonces.VerificationNonce.PubNonce,
+	})
+	if err != nil {
+		return nil
+	}
+
+	return nil
+}
+
+// taprootKeyspendSighash generates the sighash for a taproot key spend. As
+// this is a musig2 channel output, the keyspend is the only path we can take.
+func taprootKeyspendSighash(tx *wire.MsgTx, pkScript []byte,
+	value int64) ([]byte, error) {
+
+	prevOutputFetcher := txscript.NewCannedPrevOutputFetcher(
+		pkScript, value,
+	)
+
+	sigHashes := txscript.NewTxSigHashes(tx, prevOutputFetcher)
+
+	return txscript.CalcTaprootSignatureHash(
+		sigHashes, txscript.SigHashDefault, tx, 0, prevOutputFetcher,
+	)
+}
+
+// SignCommit signs the passed commitment w/ the current signing (relative
+// remote) nonce. Given nonces should only ever be used once, once the method
+// returns a new nonce is returned, w/ the existing nonce blanked out.
+func (m *MusigSession) SignCommit(tx *wire.MsgTx) (*MusigPartialSig, error) {
+	// If we already have a session, then we don't need to finalize as this
+	// was done up front (symmetric nonce case, like for co-op close).
+	if m.session == nil {
+		// Before we can sign a new commitment, we'll need to generate
+		// a fresh nonce that'll be sent along side our signature. With
+		// the nonce in hand, we can finalize the session.
+		//
+		// TODO(roasbeef): can also pass in stuff like the sighash to
+		// further bind context, etc, etc.
+		signingNonce, err := musig2.GenNonces(
+			musig2.WithPublicKey(m.localKey.PubKey),
+		)
+		if err != nil {
+			return nil, err
+		}
+		if err := m.FinalizeSession(*signingNonce); err != nil {
+			return nil, err
+		}
+	}
+
+	// Once we sign with a nonce, we'll never use it again, so it's safe to
+	// go ahead and clean up the session right here.
+	// defer m.signer.MuSig2Cleanup(m.session.SessionID)
+	//
+	// TODO(roasbeef): can't clean up here as need to combine sig
```

Comment:
perhaps worth adding a `ZeroNonces` method to `m.Signer` that can be called here?